sceKernelAddTimerEvent implementation

Fire-Cube opened the pull request:
sceKernelAddTimerEvent was implemented based on RE and the existing functions.
It’s used in CUSA05786, but I can’t confirm that it changes anything.
However, based on the RE of how sceKernelAddTimerEvent is used in the game and the output in the emulator, it appears to behave correctly.

Fire-Cube committed:
implementation

Fire-Cube committed:
add backend (WIP)

squidbus commented on src/core/libraries/kernel/equeue.h:
```diff
@@ -152,10 +152,12 @@ class EqueueInternal {
-private:
-    std::string m_name;
     std::mutex m_mutex;
     std::vector<EqueueEvent> m_events;
```
If you need to use these for your implementation, you should make a function inside the class instead of just exposing them.

Fire-Cube committed:
now should be good
- fix implementation based on homebrew tests
- demote log to debug
- make squidbus happy (hopefully)

Fire-Cube committed:
fix moved m_name

Fire-Cube committed:
fix clang

Fire-Cube committed:
replace existing event when its same id and filter

squidbus commented on src/core/libraries/kernel/equeue.cpp:
```diff
@@ -300,6 +310,83 @@ int PS4_SYSV_ABI sceKernelDeleteHRTimerEvent(SceKernelEqueue eq, int id) {
+            TimerCallback(ec, eq, event_data, interval_ms);
+        });
+
+    if (!eq->AddEvent(event)) {
```
Question: Should we add the event before creating the async timer and thread, or is there a reason not to? Thinking about the (probably very unlikely) edge case where the timer callback fires before we add the event.

Fire-Cube replied:
You're right. Additionally, this way the timer only runs if the AddEvent was actually executed correctly.

squidbus commented on src/core/libraries/kernel/equeue.cpp:
```diff
@@ -300,6 +310,83 @@ int PS4_SYSV_ABI sceKernelDeleteHRTimerEvent(SceKernelEqueue eq, int id) {
+        io_context, std::chrono::milliseconds(interval_ms));
+
+    KernelSignalRequest();
+    std::thread([]() { io_context.run(); }).detach();
```
Why do we do this? KernelSignalRequest already triggers an io_context run in the Kernel Service Thread.

Fire-Cube committed:
run timercallback after addEvent and remove useless code

Fire-Cube committed:
move KernelSignalRequest to the end

Fire-Cube committed:
clang (i hate you)
